samples/shuffles/shuffles.cc

```diff
@@ -0,0 +1,104 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+/*
+ * Test cases for the shuffle variations.
+ */
+
+#include <cstdlib>
+#include <stdio.h>
+#include <iostream>
+#include <numeric>
+
+#include "hip/hip_runtime.h"
+
+__global__ void shfl_down_int(int *In, int *Out, int Delta, int Width) {
+  Out[threadIdx.x] = __shfl_down(In[threadIdx.x], Delta, Width);
+}
+
+__global__ void shfl_up_int(int *In, int *Out, int Delta, int Width) {
+  Out[threadIdx.x] = __shfl_up(In[threadIdx.x], Delta, Width);
+}
+
+__global__ void shfl_int(int *In, int *Out, int Lane, int Width) {
+  Out[threadIdx.x] = __shfl(In[threadIdx.x], Lane, Width);
+}
+
+__global__ void shfl_down_float(float *In, float *Out, int Delta, int Width) {
+  Out[threadIdx.x] = __shfl_down(In[threadIdx.x], Delta, Width);
+}
+
+__global__ void shfl_up_float(float *In, float *Out, int Delta, int Width) {
+  Out[threadIdx.x] = __shfl_up(In[threadIdx.x], Delta, Width);
+}
+
+__global__ void shfl_float(float *In, float *Out, int Lane, int Width) {
+  Out[threadIdx.x] = __shfl(In[threadIdx.x], Lane, Width);
+}
+
+#define LAUNCH_CASE_T(TYPE, KERNEL, DELTA, WIDTH)                              \
+  std::iota(Input, Input + Threads, (TYPE)0);                                  \
+  std::fill(Output, Output + Threads, (TYPE)0);                                \
+                                                                               \
+  hipLaunchKernelGGL(KERNEL##_##TYPE, dim3(DEFAULT_WARP_SIZE), dim3(Threads),  \
+                     0, 0, (TYPE *)Input, (TYPE *)Output, DELTA, WIDTH);       \
+  hipStreamSynchronize(0);                                                     \
+                                                                               \
+  std::cout << #KERNEL "_" #TYPE << " arg=" << DELTA << " width=" << WIDTH     \
+            << ": " << std::endl;                                              \
+  for (int i = 0; i < Threads;) {                                              \
+    for (int wi = 0; wi < WIDTH; ++wi, ++i)                                    \
+      std::cout << (TYPE)Output[i] << ' ';                                     \
+    std::cout << std::endl;                                                    \
+  }                                                                            \
+  std::cout << std::endl
+
+#define LAUNCH_CASE(KERNEL, DELTA, WIDTH)                                      \
+  LAUNCH_CASE_T(int, KERNEL, DELTA, WIDTH);                                    \
+  LAUNCH_CASE_T(float, KERNEL, DELTA, WIDTH)
+
+int main(int argc, char *argv[]) {
+
+  uint Threads = DEFAULT_WARP_SIZE * 2;
```

Comment:
Shouldn't the warp size be queried from the device?

Reply:
We cannot do that as the SPIR-Vs are warp-size specific, which are produced at compile time. To make it adapt to the device at runtime we either should introduce a builtin to query the warp size also at SPIR-V runtime or fix the warp size at CHIP-SPV build-time and make the devices adapt to whatever we have fixed it to. It's not very feasible approach as we rely on kernel annotation currently to make the warp-primitives work.